Source/WebCore/Modules/url-pattern/URLPattern.h

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <wtf/Forward.h>
+#include <wtf/Ref.h>
+#include <wtf/RefCounted.h>
+#include <wtf/RefPtr.h>
+#include <wtf/TZoneMalloc.h>
+
+namespace WebCore {
+
+struct URLPatternInit;
+struct URLPatternOptions;
+struct URLPatternResult;
+
+class URLPattern final : public RefCounted<URLPattern> {
+    WTF_MAKE_TZONE_OR_ISO_ALLOCATED(URLPattern);
+public:
+    using URLPatternInput = std::variant<String, URLPatternInit>;
+    using URLPatternCompatible = std::variant<String, URLPatternInit, URLPattern>;
+
+    static Ref<URLPattern> create(const URLPatternInput, const String&& baseURL, URLPatternOptions&&);
+    static Ref<URLPattern> create(std::optional<URLPatternInput>, URLPatternOptions&&);
+    ~URLPattern();
+
+    ExceptionOr<bool> test(std::optional<URLPatternInput>, std::optional<String>&& baseURL);
+
+    std::optional<URLPatternResult> exec(std::optional<URLPatternInput>&&, std::optional<String>&& baseURL);
```

Comment:
ditto.

Reply:
Seems odd that test() / exec() functions would modify the class.

Reply:
Yes you are right! Sorry, I misread the spec. 

For these two functions,  URLPattern instance is being compared against another input. So that particular instance's properties won't change.

Reply:
(
I don't know why I thought this but my original thinking was that when the input string is passed in, URLPattern will try to parse/compare the input string with the components, eg. protocol, password, etc and that would somehow affect the instance's member properties. 

But the member properties will most likely stay the same and the input string components are probably in some temporary variable. )